### ISSUE TITLE
Image Patching -> Image Reconstructing Workflow

### ISSUE THREAD

**Issue by EtagiBI** [karolzak/keras-unet#39]
Hello,

I'm trying to implement image patching -> image reconstructing tecnhique. There're two helpful functions in the repository: `get_patches` and `reconstruct_from_patches`. On the whole the algorithm is quite clear: 
1) get a bigger image,
2) crop it to desired patches
3) feed patches to Unet as inputs
4) reconstruct predictions to a bigger mask

There're some aspects I'd like to clarify though. How am I supposed to create batches from source images of various sizes?

For instance, my batch size is 8. My input size is 512x512. I take the first two images from the set: one image is 1024x1024 and another one is 2048 x 2048. After cropping I'll get 4 patches from the first image and 16 patches from the second image. Thus, the first batch will be 4 patches from image one and 4 patches from image two. It's clear that Keras won't reconstruct the second image from just 4 pieces. 

Does it mean that I should have masks for patches, not for images?

**Comment by karolzak** [comment]
Hi @EtagiBI !
Sorry for late response.
In case you're still having problems with this I think the easiest way is to create patches as a preprocessing step rather than generating them "online" during training. So take all your images and create patches and store them in folder before training - then you can shuffle them and load into your training, add augmentations and stuff.

>Does it mean that I should have masks for patches, not for images?

Yes, in this case you need to cut your "big" masks into smaller patches as well.
Usually in such scenarios I simply create patches by adding index number at the end of filename like below:
Assume "image0.png" is my original image with 1024x1024 size and I want to generate patches 512x512 without any overlap then I would create 4 images 512x512 with following names:
image0_0.png
image0_1.png
image0_2.png
image0_3.png

It should be easy to recreate the original 1024x1024 image based on this naming convention.

**Comment by EtagiBI** [comment]
Thanks for your reply!
*Cheers* All my guesses were correct! 
I implemented this technique a couple of weeks ago, but accuracy metrics were significantly worse because of poor predictions on the patches' boundaries.

**Comment by karolzak** [comment]
>I implemented this technique a couple of weeks ago, but accuracy metrics were significantly worse because of poor predictions on the patches' boundaries.

Ah yes, this can happen - usually, using overlapping patches solves this problem. Also in case you're using `vanilla_unet` - it's known for poor performance on image borders (because of using `valid` padding for conv layers) that's why there are technics such as border mirroring to mitigate that. Original U-Net paper is full of it in case you haven't read it
